### ISSUE TITLE
Angular Elements - status + roadmap

### ISSUE THREAD

**Issue by MickL** [angular/angular#20891]
## I'm submitting a...
<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[ ] Feature request
[X] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Description
Angular Elements sounds pretty awesome. I would love to know what the status is, what the roadmap looks like and when to expect a pre-release and a release candidate.

Latest news we(the community) got was from november 2017 on Angular Connect by @robwormald 

<img width="772" alt="bildschirmfoto 2017-12-08 um 14 43 53" src="https://user-images.githubusercontent.com/5103210/33768478-78034602-dc26-11e7-8ecd-3057f6043340.png">

Also what about browser support?

## Other
In addition it would be awesome to let the community know about the status and roadmap of other upcoming stuff, too. For example i know from Angular Connect that @ocombe is working full time on improved i18n. Also @filipesilva (?) is working on Angular libraries which seem to be highly requested, too. It would be awesome for the community to have a page on angular.io(or GitHub) where to see the current backlog/roadmap of new features, the status, the effort put in, and maybe some vague dates when to expect a release candidate.

**Comment by jenniferfell** [comment]
@robwormald Per our DM, assigned to you to consider. Thanks.

**Comment by lijumathews** [comment]
Waiting to use this feature too

**Comment by pstrh** [comment]
We hope to implement a micro frontend architecture with Angular Elements, e.g. combine several Angular-based elements/application in one page. Therefore the question: will it be possible to combine several Angular versions on one page with Angular elements? It would be great if this question and the possible impact are addressed in the documenation.

**Comment by skreborn** [comment]
@pstrh I would assume that since Angular Elements creates a sandbox for itself and is bound to its host element, you would be able to use different Angular versions just like you could use any other framework for the other elements on your page. Different elements shouldn't need to know what code controls the others.

**Comment by baldwindc** [comment]
+1 that a general release date / roadmap would be helpful.

**Comment by mlc-mlapis** [comment]
@skreborn ... can you describe your idea how it would be possible to mix different versions and keep also a reasonable size of the whole?

**Comment by baldwindc** [comment]
For a playground to learn Angualr Elements, how up to date is https://github.com/robwormald/angular-elements?

I'm having trouble finding many articles or resources about angular-elements other than "Here's what we want it to do." It's intended purpose is exactly what my organization is in the beginning phases of doing at the moment. I'm currently attempting to convert our existing angular 1.6 app over to use https://github.com/CanopyTax/single-spa.

- Is angular-elements being developed internally (or is there an open-source alpha build somewhere I can examine)?
- (expounding from my previous comment): When can we expect to see a beta release?
- Will this allow older version of angular to be built incorporated into the angular-elements "microframwork"? 
-

**Comment by mlc-mlapis** [comment]
@skreborn ... ok, I just wanted to confirm if you have any `background` idea that would change the game. 😄 My seeing is that it is a general limitation ... and I would be surprised a lot if any `extra` solution will appear. Actually I think that if you want to use several components with a minimum footprint you have to do a similar step as CLI `tree-shaking elimination` does today ... to generate a minimal shared lib which would contain only the absolutely necessary parts from standard Angular libs ... so it is logical that using of the same Angular version is a must.

**Comment by robwormald** [comment]
Updates on this: we're targeting v6 for an initial, experimental release of elements. We'll be validating it with Angular.io, and aim to solve that use case first. 

We'll get more into details as we get closer to release. Feel free to ask questions in here - can't promise I'll answer them all, but we are watching and capturing use cases. 

Conceptually, Elements (or at least, the initial implementation) is pretty straightforward, and doesn't really change too much about Angular. In parallel, there's major work happening on the view engine, and our early numbers are extremely promising in terms of code size. Again, v6 is mainly *foundational* work - post v6 will see some very interesting things happen that we'll talk more about soon. 

Please keep in mind this is a **Labs** project, and as such we make no guarantees on ship dates, functionality, etc. We know plenty of people are excited about this (as are we!), and our aim is to get something useable into devs hands sooner than later, so we can learn and iterate - but understand that things may not be as batteries-included as stable angular APIs. Support in CLI and such will come, but you may have to roll your own webpack config (or similar) in the short term.

Stay tuned!

> We hope to implement a micro frontend architecture with Angular Elements, e.g. combine several Angular-based elements/application in one page. Therefore the question: will it be possible to combine several Angular versions on one page with Angular elements? It would be great if this question and the possible impact are addressed in the documenation.

In theory, yes, but this is not ideal. This is part of the experimental process, but (along with upcoming view engine changes) I would hope that running multiple versions would be unnecessary.

> @pstrh I would assume that since Angular Elements creates a sandbox for itself and is bound to its host element, you would be able to use different Angular versions just like you could use any other framework for the other elements on your page. Different elements shouldn't need to know what code controls the others.

This is conceptually correct indeed, but again, not ideal. 

> For a playground to learn Angualr Elements, how up to date is https://github.com/robwormald/angular-elements?

It's not - it was purely a proof of concept. It's a reasonably similar approach - the thing to understand about it is that it's not anything magical - it would be possible for any developer to write a similar layer (so if you need something like this RIGHT NOW, it would be worth reading and understanding...) - it doesn't require any support from angular's compiler. 

The current codebase is on the labs/elements branch - https://github.com/angular/angular/tree/labs/elements/packages/elements/ but we expect to merge it to master shortly, pending initial docs.

> m having trouble finding many articles or resources about angular-elements other than "Here's what we want it to do." It's intended purpose is exactly what my organization is in the beginning phases of doing at the moment. I'm currently attempting to convert our existing angular 1.6 app over to use https://github.com/CanopyTax/single-spa.

This is why this falls under Angular Labs - we want to be open about what we're doing, but set the expectation this is not production ready yet. We'll have some basic docs for the initial release, and full documentation and integration when it graduates from Labs status.

> Is angular-elements being developed internally (or is there an open-source alpha build somewhere I can examine)?
> (expounding from my previous comment): When can we expect to see a beta release?

See above

> Will this allow older versions of angular to be built incorporated into the angular-elements "microframwork"?

I'm not sure I understand this question - it's unlikely we'll do any backwards facing work on Elements, but again, there's nothing "special" about elements, so if it's something you'd require, it would likely be possible to do yourself. Hint: Angular Elements and NgUpgrade (specifically, downgrade) are basically doing the same thing - bridging Angular's API to the Custom Elements spec and AngularJS APIs, respectively. 

> @mlc-mlapis I have been thinking about the size implications myself, and I have honestly no idea how this could be done while also keeping reasonable constraints on bundle sizes. With each version that you use, you'll have to load the entire bundle separately. I didn't say that it would be a good idea to mix different angular versions, only that I suppose that they could theoretically work side-by-side without interfering with one another.

I'm not going to quote any numbers, but suffice to say, we're reasonably confident that within the next couple major versions of Angular, bundle size should be the least of your concerns. More on that later :)

**Comment by alshdavid** [comment]
Holding my breath for this. Looking for updates on the progress every second day.

My team want to move to Vue from Angular 1.6. My hope is that the ability to integrate Angular Elements into the existing codebase (easing the transition to Angular) will shift the decision in favour of Angular over Vue.

If the release of v6 is reported to be around March/April and you're planning some progress around that time, then at least I know how much time I need to buy. 😄

**Comment by MickL** [comment]
I think Elements could really really push Angular alot. But we also need the ability to create libraries easy as soon as possible. Anyway the main problem is that the community is not getting updated. We dont know what you guys are working on and when to expect something but it really is import for our projects to plan. @robwormald @filipesilva @ocombe

**Comment by ocombe** [comment]
If you look into the milestone for Angular v6: https://github.com/angular/angular/milestone/81
There's a task for Angular Elements MVP: https://github.com/angular/angular/issues/21573 that you can follow.

**Comment by MickL** [comment]
Ok thank you. I am just saying that NOW we got informations on Elements and i18n but we had to ask for it for long time. And also they are in the middle of long issues. It would be great to have a place where see this, what to expect from Angular in the future, what is currently in work, what has priority etc etc. E.g. a summirization of you presentation on Angular Connect + the updates in the issue. Same goes for Elements and libraries. And this are only the three things i know i bet there is much more going on.

**Comment by ocombe** [comment]
I understand your concern, and I've asked the same question last week to the team. The reply was that when you make announcements, even if you're being cautious, people will take what you say for a fact. If it works out as you said, it's nice and everyone is happy, but that's rarely the case in development, things get postponed because other things take the priority, or things get canceled because it didn't work or issues arise that you didn't anticipate, etc...
The point is that even if you say that it's an estimation, if it changes some people will get angry because they took what you said for granted and planed accordingly (for a project, for a business, etc...).
Let's say that we announce "Angular Elements will be released in March", and you had convinced your boss to use this for the next project instead of Vue (like @alshdavid wants to do), and it turns out that in March it's not ready, it's buggy, or we had to postponed because we need to finish something else first. Then your boss comes to you and blames you, and in return you'll blame the Angular team.

At Angular Connect, I was so sure that we would get runtime i18n and code translations done for 5.3 that I said that during my talk. We were making good progress, we had design docs and we were working on it already so I had high hopes.
It turns out that some core changes (new Renderer) will make our work much easier, but also obsolete if we had finished that part before. So we decided to wait a little and work on runtime i18n with the new renderer. It postponed that to v6 instead of 5.3, but it's for the best.
Now apparently no one has blamed me for the delay this time, but it could have happened and I would have felt terrible (even if it was not really my fault and it was for the best).

That's why we cannot make announcements like that. From what I've heard, we will try our best in the future to share our roadmap and we work on, and probably that hiring a new devrel will also help for the communication (Rob & Stephen already have a lot of ground to cover).

I hope that you understand our position as well. As a developer I really really want to share some of the new and exciting stuff that we have planned, and to give you actual numbers for the new bundle size, but I cannot in good conscience do that for now.

**Comment by johnfabian** [comment]
Will Angular Elements even the experimental release version work with SharePoint 2016 and SPFX, with multiple instances on the page? This is the most important feature my team needs!

**Comment by sandangel** [comment]
@johnfabian does it work with vanilla js and html? if it does, angular element will work too

**Comment by jared-christensen** [comment]
This might be a crazy question, but could I use an angular element inside of another angular app? I tried it but I get an error. "Uncaught Error: Zone already loaded."

**Comment by sandangel** [comment]
@jared-christensen could you share your demo?

**Comment by artaommahe** [comment]
What about cross-elements/elements-mainapp interactions? Common services like auth/user e.g. Any proposal/plans here? Currently it looks like any used widget should have his own copy of necessary common services or receive data from top-element inputs (which does not eliminate the need of duplicated services for their methods usage)

**Comment by skreborn** [comment]
@artaommahe I don't think there will be any sharing between them. Angular Elements are sand-boxed by design. The implementation outside doesn't even know that the element is powered by Angular, or the other way around. At least that's what I gathered.

**Comment by jared-christensen** [comment]
@manfredsteyer Thanks for the tip, I ended up just tuning zone off altogether for the elements. This article was helpful. Was very easy to turn off. https://blog.angularindepth.com/do-you-still-think-that-ngzone-zone-js-is-required-for-change-detection-in-angular-16f7a575afef

**Comment by trotyl** [comment]
@artaommahe What's your definition of "main app"? There can be as many application instances as your want without using Angular Elements, there will be **NO** state sharing between them.

In current WIP implementation, you can already [passing an `Injector`](createNgElementConstructor) when defining Elements (via `createNgElementConstructor`), so it can be made part of any application if you want.

**Comment by artaommahe** [comment]
> What's your definition of "main app"? There can be as many application instances

any of them or some common part with common services. I'm talking about widgets that are truly separated from bigger or common app building and development pipeline. Our medium-sized app currently builds more than 7 mins and its not cool. There are a lot of parts that can be developed separately but require some common service from main app.

> you can already passing an Injector when defining Elements

but i cant get proper instance by class from this injector cause app/widgets are builded separately (only if inject it by string and have a copy of all common services interfaces)

**Comment by trotyl** [comment]
> but i cant get proper instance by class from this injector cause app/widgets are built separately

@artaommahe How would you like to share the injectables without same token? Technically?

**Comment by artaommahe** [comment]
> How would you like to share the injectables without same token? Technically?

it's one of the problems ) looks like only via string alias like `@Inject("smthService") smthService: ISmthService`

**Comment by robwormald** [comment]
> Will Angular Elements even the experimental release version work with SharePoint 2016 and SPFX, with multiple instances on the page? This is the most important feature my team needs!

> This might be a crazy question, but could I use an angular element inside of another angular app? I tried it but I get an error. "Uncaught Error: Zone already loaded."

Yes, BUT, you should likely build such things without Zones - they're the major issue for environments like Sharepoint - so if its something you want immediately, I'd suggest not including Zones, turning on the `noop` Zone, and handling change detection yourself:
```
platformBrowser.bootstrapModuleFactory(Whatever, { ngZone: 'noop' });
```

In general, this model is something we'll make more ergonomic in the future (eg, make the `async` pipe actually be a `push` pipe), but you can just inject and call `changeDetectorRef.detectChanges()` yourself. If you're using Observables / NgRx / similar, you're likely 90% of the way there already.

> What about cross-elements/elements-main_app interactions? Common services like auth/user e.g. Any proposal/plan here? Currently it looks like any using widget should have his own copy of necessary common services or receive data from top-element inputs (which does not eliminate the need of duplicated services for their methods usage).

> Variety of questions about injector/DI stuff
 
Elements use the same DI / Injection mechanism as the rest of Angular, and provides two mechanisms - you can (must, currently) provide a (common) injector when the Element is registered:
```
createCustomElement(SomeComponent, { injector: someInjector });
```
or override that for a specific instance imperatively:
```
const XFoo = customElements.get('x-foo');

const instance = new XFoo(someOtherInjector);
```
This makes a lot of sense in an app like angular.io, where we're using Elements "inside" of an Angular application. It doesn't make as much sense for standalone "widgets". 

The main use case we're supporting in the initial V6 release is the angular.io case, and so passing in the injector is relatively natural. Post v6, as we move into Ivy and Elements, injectors should become optional (or "implicit").

Also keep in mind that depending on a specific injector at the instance level (vs the registration level) means that you will not be able to declaratively instantiate elements, as you don't get an opportunity to pass an injector in if you do 
```
document.createElement('x-foo')
```
or 
```
<x-foo></x-foo>
```

**Comment by trotyl** [comment]
@robwormald The scenario described by @artaommahe is about bring several copies of `@angular/core` (may even be difference versions) at runtime, so there'll not even be a shared root scope. I don't think passing an Injector would work in this case, neither will Ivy.

**Comment by alshdavid** [comment]
Are there plans to support the use of `document.createElement()`?

**Comment by robwormald** [comment]
@alshdavid ? Yes, that’s already supported.

**Comment by studioTeaTwo** [comment]
I see that a method `renderComponent` provided by ivy is imported and referenced at component, not module.
I think it's good that a method to `elementify component` can also be imported and used at per component.

**Comment by jared-christensen** [comment]
I created two apps using the angular cli and turned the app module of each into angular elements. When I tried to place them on the page I noticed I have to order the bundle scripts like this to get them to work without errors.

1. inline (element app 1)
2. polyfills (element app 1)
3. main ( element app 1)
4. inline (element app 2)
5. main ( element app 2)

Notice I removed the polyfills for the 2nd elements. I also noticed I have to build them with the --prod flag or I can not get them to work.

Do I have the right idea here? Can we have elements from separate apps on the same page, I would assume so? Should I be building them with the angular cli or will they need a different build process? We would likely have multiple teams building elements to be used on the same page, how can we manage the polyfills file? Just curious what the plans might be for building and deploying elements.

**Comment by soates** [comment]
How will Elements work with different version of Angular? Say I use a component that has been created by version 6.0 and in the future another component is created by version 7.0?

**Comment by trotyl** [comment]
@soates The question is meaningless, you can already do that today, why would Angular Elements prevent you from doing that deliberately?

**Comment by MickL** [comment]
As far is I understand an Angular Element is an complete molecule. This means you will have a js-file for each Element with no dependencies that works in every environment.

**Comment by robwormald** [comment]
> I think it's good that a method to elementify component to custom element can also be imported and used at per component.
@studioTeaTwo that's basically how it will work in V7, or potentially, we'll just have the compiler do that step.

> I created two apps using the angular cli and turned the app module of each into angular elements. When I tried to place them on the page I noticed I have to order the bundle scripts like this to get them to work without errors.
@jared-christensen understand that the v6 version of elements is not for the case you're describing. it's possible, but you're not going to get a magical CLI build that does it right now. if this is something you want to explore, i'd suggest doing the bundling yourself (via rollup etc). This case will be more automatic in v7 with Ivy, when the code size isn't really an issue.

> @soates The question is meaningless, you can already do that today, why would Angular Elements prevent you from doing so deliberately?
@trotyl be nice. this is new to lots of people.

> As far is I understand an Angular Element is an complete molecule. This means you will have a js-file for each Element with no dependencies that works in every environment.

This isn't the right mental model - Angular Elements are still Angular under the hood. How they're bundled / packaged / deployed is flexible for whatever case, but they will (and likely forever will) have some level of dependency on Angular. That might mean bundling the subset of Angular core you use into each standalone element, or it might mean sharing Angular via a script tag (like AngularJS).

**Comment by MickL** [comment]
Very interesting, makes me excited! 

Do you think, with Ivy, bundling a single element with Angular will be comparable, in size, to handwritten code? E.g. a simple datepicker with everything (Angular) included at <5kb gzipped?

**Comment by johnfabian** [comment]
@robwormald  I have angular working in sharepoint without zones and without angular elements and also with multiple webparts on the page using:

```
platformBrowserDynamic().bootstrapModuleFactory(
        AppModuleNgFactory, { ngZone: 'noop' })
        .catch(
            err => console.log(err));
```

I'm also calling changeDetectorRef.detectChanges() but my issue now is I'm using 3rd party form controls  and they are not working, they are implement ControlValueAccessor, the write value looks like below

```
writeValue(value: any) : void {   
      this.value = value;
      this.cd.markForCheck();      
  }
```

The inital value never gets set. I can modify writeValue and change markForCheck to detectChanges but seems I would have to do this for every control.   Is this the correct approach? 

Will Angular  Elements solve this issue so I can use zones and not have to modify the controls? Are there any tutorials on using Angular Elements? I would like to test the 3rd party controls inside of Angular Elements but not sure how to start with it yet.

**Comment by johnfabian** [comment]
How do I include the polyfills in an SPFX webpart? I tried adding in external in config.json, no luck,  tried just importing into the webpart.ts file, not luck. My Component works great in chrome but we also have users using IE 11 as well but not sure how this works with the spfx webparts

**Comment by johnfabian** [comment]
Will I still have to have zones turned off to use Angular Elements with SharePoint SPFX? If so, I don't see this being useful for SharePoint. 

Reason being is I'm using 3rd party controls and so far for every control I've used I've had to go into the source and add detectChanges() all over the place in order for the controls to work with { ngZone: 'noop' } zones turned off.  

Another issue is I can't get anything to render in IE and have no clue how to get the polyfills to work in SPFX, I've tried adding the js files to Externals, tried importing into the webpart.ts files, nothing renders in IE which is still popular with gov clients unfortunately. 

I'm running out of time on my project and spend so much time fighting the framework and now have 2 sets of controls, One set for full page apps and another set of controls to work with zones turned off for SPFX with detectChanges() all over the place so hoping Elements will help solve this issue.

Can anyone please help with getting SPFX and Angular 5 to render in IE?

**Comment by robwormald** [comment]
@johnfabian questions specific to how to do ____ in spfx are best asked in one of the SPFX repos - we don't have a huge amount of experience with it. Feel free to cc me over there and I'm happy to talk through it on their end. 

Re: Zones, this is why things are experimental :) Zones and SPFX can theoretically play together nicely, but you need to be sure you only have one copy of Zone.js (like you should only have one copy of the various polyfills...)

**Comment by robwormald** [comment]
@MickL it's very, very hard to give accurate estimates (and not get quoted all over the internet forever...) but that's a reasonably accurate estimation, yes. Simple TodoMVC apps with a few components are around 13kb right now, and that's with a lot of legacy stuff still in the bundle.

**Comment by gogakoreli** [comment]
@robwormald Hey Rob, I just tried to reuse Angular Elements inside another Angular project and my problem is that I cant make property binding to work on custom element imported from somewhere else. I think that I don't know something important. Is it even possible?

**Comment by artaommahe** [comment]
What about content projection and injecting parent element for custom elements? Smth like tabs
```
<tabs>
  <tab>smth</tab>
  <tab>smth2</tab>
</tabs>
```
i'm about injecting TabsComponent component inside TabComponent

**Comment by robwormald** [comment]
@jur4ik54 elements are able to use a shared injector (that is, you could share a service between them). 

@gogakoreli inside an Angular template, we enforce typechecking. You can do `schemas: [ CUSTOM_ELEMENTS_SCHEMA` in the NgModule you want to use the CE in.

@artaommahe `ng-content` works in a limited fashion (though `@ContentChild(ren)` cannot), or alternately, you can use the native `<slot>` API for content projection.

**Comment by gogakoreli** [comment]
Thank you for info and actually I was talking about one way and two way property binding on the Angular Custom Element imported inside my project. Is it possible to use bindings?

**Comment by artaommahe** [comment]
@robwormald so no parent injection in my case now? Is this and ContentChild support planned?

**Comment by DaSchTour** [comment]
@robwormald I think it's important to create comprehensive documentation for all this topic. The current documentation about angular elements is very basic and doesn't tell about any of this restrictions. I think there is a great danger, that angular elements will raise high expectations and will disappoint people because of bad documentation. There are many people that will not that didn't dive that deep into this topic and read more than just the documentation.

**Comment by Rainson12** [comment]
@robwormald I am following a few guides to use angular elements with spfx. Sadly they all use the old method: 

`registerAsCustomElements(
   [ HelloWorld ],
   () => platformBrowser().bootstrapModuleFactory(HelloWorldModuleNgFactory, { ngZone: 'noop' })
 ).catch(err => console.log(err));`


As registerAsCustomElements doesnt exist i wonder about how this should be translated? It previously was pasted directly into a ts file without an exporting class / constructor.

**Comment by idibidiart** [comment]
Does anyone have any feedback on this demo and its limitations if any?

https://github.com/wishtack/wt-angular-elements-demo

I'm looking to use this pattern for a complex Angular component that will be shared as a Custom Element and used in React, Ember and other places. Whether it'll be 200K or 300K in size it doesn't matter to me because that component takes up most of the page anyway, and is almost an app in itself.

To make things more challenging, I am completely new to Angular, a refugee/convert from React land.

Cc: @robwormald

**Comment by jenniferfell** [comment]
@jbogarthyde As we assess our backlog, would you please review this and work with @robwormald and @andrewseguin to write up an elements doc backlog? We can talk more offline. Thanks!

**Comment by pkerblaze** [comment]
To inform, I had the same problem with `Uncaught Error: Zone already loaded.` but the custom element was working. When I removed the polyfill.ts from the build process (see: https://angularfirebase.com/lessons/angular-elements-quick-start-guide/ and https://github.com/angular/angular/issues/23732#issuecomment-387915011) I got a new error  `Uncaught Error: In this configuration Angular requires Zone.js` and the custom element was no more working.

This use case is when you load the script inside the DOM 
```
<body>
  <app-root></app-root>
  <script src="assets/custom-element.js"></script>
</body>
```

But if you remove the script from the DOM and insert it into the consumer building time (angular.json) _without polyfills.ts during the build process of the custom element_, it work without error (didn't try for multiple custom elements) : 
```
"projects": {
    "custom-element-consumer": {
      // ..
      "architect": {
        "build": {
            // ...
            "scripts": [
               "src/assets/custom-element.js"
            ]
            // ...
```

So you can use a Custom element inside another angular app, but you need to add it at build time and not at runtime for the moment, hope it'll help people who want to try it or if someone find a way to add it at runtime without error !

**Comment by DaSchTour** [comment]
I'm really more interested in the other way. How to get a small JS for a webcomponent out of angular.
I really miss some more documentation and support from `@angular/cli`.
I made a proposal how I would expect it to work on angular/cli repso https://github.com/angular/angular-cli/issues/8988

**Comment by MickL** [comment]
I guess this will be possible and documentated not before Angular 7 and the Ivy renderer.

**Comment by DaSchTour** [comment]
That's really bad and frustrating as I'm losing traction on this. Angular has so many great features but I can't sell the internally because they are documented very poorly. And recently I also heard the argument, that projects like to use Vue over Angular because Vue has better documentation. I really hope to get additional arguments why we should use Angular over Vue but it get's harder and harder with this kind of incomplete documentation.

**Comment by MickL** [comment]
We are talking about future proposals. Personally i would prefer them to work on the new features instead of on documentation for something thats not even finished. 
 
I agree that we need more informations on what they are working on and planning and maybe when to expect something.

There are tons of arguments for Angular and TypeScript but you should discuss this somewhere else.

**Comment by MickL** [comment]
Im closing this issue since it is no todo that can be completed. Discussions and status updates from rob are welcome!

**Comment by SimmeNilsson** [comment]
What would be the recommended way to listen for attribute or property changes on an angular element from a non-angular outside world? Should mutation observer work?

**Comment by MickL** [comment]
Shouldnt `.addEventListener('myOutput', fn)` work? Or jQuery `.on()` or RxJs `fromEvent()`?

**Comment by SimmeNilsson** [comment]
If so then I must be doing something wrong (or missed something). Should that event trigger as soon as I change a property or attribute?
Looking for something like Vue's $watch. For example, if element has would have a dirty flag I would like to subscribe to state change of that one.

**Comment by DaSchTour** [comment]
The question is how angular is doing the attribute changes internally. If they use [setAttribute](https://developer.mozilla.org/en-US/docs/Web/API/Element/setAttribute) then MutationObserver should work. Inside the Component you have the onChanges hook that works somehow similar to MutationObserver but from the inside of the component.

**Comment by emilio-martinez** [comment]
Angular does what any standard built-in or custom element would do.

If the literal goal is to listen to attribute changes, then yes, you can observe DOM mutations—it's the DOM, after all! There's no Angular black magic at play—[see the DOM Render implmentation](https://github.com/angular/angular/blob/6.1.0/packages/platform-browser/src/dom/dom_renderer.ts#L152), for example. However, perhaps it's best to stick with standard DOM patterns such as events. Similarly, is this regard Angular does what you'd expect from a DOM perspective—emit `CustomEvent`s for every `Output` emit. Feel free to [check out that implementation as well](https://github.com/angular/angular/blob/6.1.0/packages/elements/src/create-custom-element.ts#L168).

**Comment by emilio-martinez** [comment]
@playground not quite sure what you mean by "everything works fine but throws the same errors", but otherwise the error you got is a fairly common custom elements error:

As a bit of background, the custom element spec (https://html.spec.whatwg.org/multipage/scripting.html#custom-element-conformance) outlines that only ES2015 classes may be passed to `customElements.define`. Of course, it's still very common to serve ES5 due to browser support, so most people will ship ES5-style custom element classes, as is likely your case.

The thing to be aware of here is that browsers that _natively_  support `customElements.define` still expect ES2015 classes and will _not_ take ES-style classes because they don't properly extend classes like HTMLElement. To work around this, you will want to use the ES5 custom elements adapter (https://github.com/webcomponents/webcomponentsjs/blob/master/README.md#custom-elements-es5-adapterjs) that the webcompoments org ships when you're compiling to ES5. That will monkey-patch the native implementation to allow for ES5-style classes in your custom elements.
